DigitalLearningSolutions.Data/Services/CryptoService.cs

```diff
@@ -1,65 +1,88 @@
-
-            if (storedPasswordIsIncorrectLength || storedPasswordHasIncorrectVersionHeader)
-            {
-                return false;
-            }
-
-            var salt = new byte[SaltSize];
-            Buffer.BlockCopy(hashedPasswordBytes, 1, salt, 0, SaltSize);
-
-            var storedSubkey = new byte[PBKDF2SubkeyLength];
-            Buffer.BlockCopy(hashedPasswordBytes, 1 + SaltSize, storedSubkey, 0, PBKDF2SubkeyLength);
-
-            var deriveBytes = new Rfc2898DeriveBytes(password, salt, PBKDF2IterationCount);
-            var generatedSubkey = deriveBytes.GetBytes(PBKDF2SubkeyLength);
-
-            return generatedSubkey.SequenceEqual(storedSubkey);
-        }
-
-        public string GetPasswordHash(string password)
-        {
-            var salt = new byte[SaltSize];
-            new RNGCryptoServiceProvider().GetBytes(salt);
-
-            var deriveBytes = new Rfc2898DeriveBytes(password, salt, PBKDF2IterationCount);
-            var generatedSubkey = deriveBytes.GetBytes(PBKDF2SubkeyLength);
-
-            var hashBytes = new byte[CorrectStoredPasswordLength];
-            hashBytes[0] = CorrectStoredPasswordVersionHeader;
-            Buffer.BlockCopy(salt, 0, hashBytes, 1, SaltSize);
-            Buffer.BlockCopy(generatedSubkey, 0, hashBytes, 1 + SaltSize, PBKDF2SubkeyLength);
-
-            return Convert.ToBase64String(hashBytes);
-        }
-    }
-}
+﻿namespace DigitalLearningSolutions.Data.Services
+{
+    using System;
+    using System.Linq;
+    using System.Security.Cryptography;
+    using Microsoft.Extensions.Logging;
+
+    public interface ICryptoService
+    {
+        bool VerifyHashedPassword(string? hashedPassword, string password);
+        string GetPasswordHash(string password);
+    }
+
+    public class CryptoService : ICryptoService
+    {
+        private const int SaltSize = 16;
+        private const int PBKDF2SubkeyLength = 32;
+        private const int PBKDF2IterationCount = 1000;
+        private const int CorrectStoredPasswordLength = 1 + SaltSize + PBKDF2SubkeyLength;
+        private const byte CorrectStoredPasswordVersionHeader = 0x00;
+
+        private readonly ILogger<CryptoService> logger;
+
+        public CryptoService(ILogger<CryptoService> logger)
+        {
+            this.logger = logger;
+        }
+
+        public bool VerifyHashedPassword(string? hashedPassword, string password)
+        {
+            if (string.IsNullOrWhiteSpace(hashedPassword))
+            {
+                return false;
+            }
+
+            byte[] hashedPasswordBytes;
+
+            try
+            {
+                hashedPasswordBytes = Convert.FromBase64String(hashedPassword);
+            }
+            catch (FormatException e)
+            {
+                logger.LogError(
+                    e,
+                    "Could not verify password as user's password hash was not a valid base 64 string."
+                );
+                return false;
+            }

```

Comment:
Only change was adding in the logger and this try-catch.
I'm not sure about this log. It would be nice to tell the app maintainer who has the invalid password hash, but I also feel like that's a dangerous thing to log...

This is the output of the log:
`[09:59:49 ERR] Could not verify password as user's password hash was not a valid base 64 string.
System.FormatException: The input is not a valid Base-64 string as it contains a non-base 64 character, more than two padding characters, or an illegal character among the padding characters.
   at System.Convert.FromBase64CharPtr(Char* inputPtr, Int32 inputLength)
   at System.Convert.FromBase64String(String s)
   at DigitalLearningSolutions.Data.Services.CryptoService.VerifyHashedPassword(String hashedPassword, String password) in C:\work\hee\DLSV2\DigitalLearningSolutions.Data\Services\CryptoService.cs:line 40
`

Reply:
Yeah, I think that's the only time it would happen. True, probably better if the user saw there was an error rather than being locked out of their account. Do you reckon I should re-throw here or just get rid of the log altogether?

Reply:
I'd probably get rid of it, but double check that the Exception that bubbles up in that case makes sense. 👍

Reply:
Yup, looks fairly clear. Here it is, in case you want to see it:

```System.FormatException: The input is not a valid Base-64 string as it contains a non-base 64 character, more than two padding characters, or an illegal character among the padding characters.
   at System.Convert.FromBase64CharPtr(Char* inputPtr, Int32 inputLength)
   at System.Convert.FromBase64String(String s)
   at DigitalLearningSolutions.Data.Services.CryptoService.VerifyHashedPassword(String hashedPassword, String password) in C:\work\hee\DLSV2\DigitalLearningSolutions.Data\Services\CryptoService.cs:line 28
   at DigitalLearningSolutions.Data.Services.LoginService.<>c__DisplayClass3_0.<VerifyUsers>b__0(DelegateUser du) in C:\work\hee\DLSV2\DigitalLearningSolutions.Data\Services\LoginService.cs:line 45
   at System.Linq.Enumerable.WhereListIterator`1.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at DigitalLearningSolutions.Data.Services.LoginService.VerifyUsers(String password, AdminUser unverifiedAdminUser, List`1 unverifiedDelegateUsers) in C:\work\hee\DLSV2\DigitalLearningSolutions.Data\Services\LoginService.cs:line 44
   at DigitalLearningSolutions.Web.Controllers.LoginController.Index(LoginViewModel model) in C:\work\hee\DLSV2\DigitalLearningSolutions.Web\Controllers\LoginController.cs:line 65
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Serilog.AspNetCore.RequestLoggingMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)```